src/js/fable-core/Array2.fs

```diff
@@ -1,16 +1,732 @@
+
+let zip3 (array1: 'T[]) (array2: 'U[]) (array3: 'U[]) (cons: ArrayCons) =
+    // Shorthand version: map3 (fun x y z -> x, y, z) array1 array2 array3
+    if array1.Length <> array2.Length || array2.Length <> array3.Length then failwith "Arrays had different lengths"
+    let result = cons.Create(array1.Length)
+    for i = 0 to array1.Length - 1 do
+       result.[i] <- array1.[i], array2.[i], array3.[i]
+    result
+
+let chunkBySize (chunkSize: int) (array: 'T[]): 'T[][] =
+    if chunkSize < 1 then invalidArg "size" "The input must be positive."
+
+    if array.Length = 0 then [| [||] |]
+    else
+        let result: 'T[][] = [||]
+        // add each chunk to the result
+        for x = 0 to System.Math.Floor(float(array.Length) / float(chunkSize)) |> int do
+            let start = x * chunkSize;
+            let end' = start + chunkSize;
+            let slice = slice array start end'
+            push result slice |> ignore
+
+        result
+
+let fill (array: 'T[]) offset count value =
+    for i = offset to offset + count - 1 do
+        array.[i] <- value
+
+let splitAt (index: int) (array: 'T[]): 'T[] * 'T[] =
+    if index < 0 then invalidArg "index" LanguagePrimitives.ErrorStrings.InputMustBeNonNegativeString
+    if index > array.Length then invalidArg "index" "The input sequence has an insufficient number of elements."
+
+    slice array 0 index, sliceFrom array index
+
+let inline compareWith (comparer: 'T -> 'T -> int) (array1: 'T[]) (array2: 'T[]) =
+    let length1 = array1.Length
+    let length2 = array2.Length
+
+    let mutable i = 0
+    let mutable result = 0
+
+    if length1 < length2 then
+        while i < array1.Length && result = 0 do
+            result <- comparer array1.[i] array2.[i]
+            i <- i + 1
+    else
+        while i < array2.Length && result = 0 do
+            result <- comparer array1.[i] array2.[i]
+            i <- i + 1
+
+    if result <> 0 then result
+    elif length1 = length2 then 0
+    elif length1 < length2 then -1
+    else 1
+
+let exactlyOne (array: 'T[]) =
+    if array.Length = 1 then array.[0]
+    elif array.Length = 0 then invalidArg "array" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+    else invalidArg "array" "Input array too long"
+
+let head (array : 'T[]) =
+    if array.Length = 0 then invalidArg "array" LanguagePrimitives.ErrorStrings.InputArrayEmptyString
+    else array.[0]
+
+let tryHead (array : 'T[]) =
+    if array.Length = 0 then None
+    else Some array.[0]
+
+let tail (array : 'T[]) =
+    if array.Length = 0 then invalidArg "array" "Not enough elements"
+    slice array 1
+
+let item index (array: _[]) =
+    array.[index]
+
+let tryItem index (array:'T[]) =
+    if index < 0 || index >= array.Length then None
+    else Some array.[index]
+
+let toList array = 
+    List.ofArray array
+
+
+// TODO: Consider optimizing functions below via manual inlining
+// -or- re-implementing via native js arrays functions (i.e. reduce(...))
```

Comment:
Sometimes a `for` loop is faster than native JS array functions so your current approach should be fine. About inlining, because this code will be distributed in JS form we cannot use `inline` here, inlining would have to be done in the `Replacements` module if necessary.

Reply:
Agree! And as for "manual inlining" I meant that stuff when you unwrap/inline function by hand, and not asking the compiler via `inline` keyword. E.g. when you have `sum` implemented via 'fold` first, but then replaces `fold` with a loop/recursion code yourself.

Anyway, as I've said already in a comment above, it smells like a premature optimization to me :)